torch/nn/functional.py

```diff
@@ -4841,28 +4841,79 @@ def _in_projection(
 greater than 0.0 is specified.
 
+.. code-block:: python
+
+    # Composite Implementation
+    attn_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0) if is_causal else attn_mask
+    attn_mask = attn_mask.masked_fill(not attn_mask, -float('inf')) if attn_mask.dtype==torch.bool else attn_mask
+    attn_weight = torch.softmax((Q @ K.transpose(-2, -1) / math.sqrt(Q.size(-1))) + attn_mask, dim=-1)
+    attn_weight = torch.dropout(attn_weight, dropout_p)
+    return attn_weight @ V
+
+.. warning:: This function is beta and subject to change.
+
+Note:
+    This function calls into one of three backends:
+        * `FlashAttention: Fast and Memory-Efficient Exact Attention with IO-Awareness`_
+        * `Memory-Efficient Attention`_
+        * A pytorch implementation defined in c++ matching the above mathematical formulation
+
+    The function defaults to selecting the highest-performing implementation based on the inputs provided.
+    However, each of the fused kernels has specific input limitations.
+    If you require a specific backend to be utilized, there exists functions to enable or disable specific backends.
+    Please note that all backends are enabled by default.
+
+    For example :func:`~torch.backends.cuda.enable_flash_sdp` can be used to enable/disable FlashAttention.
+    The context manager :func:`~torch.backends.cuda.sdp_kernel` can be used to enable/disable the backends
+    for a specific scope.
+
+    If a user wants to enforce that one of the fused implementations is used, disable the math fallback
+    using one of the above mechanisms. If for some reason a fused implementation is not available,
+    the function will throw an error with the reasons why the fused implementation was not used.
+
+    The numerical accuracy of the fused kernels has been tested but due to the nature of fusing floating point operations
+    The deviations from the infinite precision implementation may be significant. If that is the case we encourage users
+    to please file an issue. A work around would be disabiling the fused kernels and using the math fallback. For more
+    information please see :doc:`/notes/numerical_accuracy`.
+
+Note:
+    {cudnn_reproducibility_note}
+""".format(**reproducibility_notes)
+    + r"""
+
 Args:
-     query (Tensor): Query tensor; shape (N, ..., L, E)
-     key (Tensor): Key tensor; shape (N, ..., S, E)
-     value (Tensor): Value tensor; shape (N, ..., S, E)
-     attn_mask (optional Tensor): Attention mask; shape (N, ..., L, S) or (L, S). Currently, only a boolean mask
-         is supported, where a value of True indicates that the element *should* take part in attention.
-     dropout_p (float): Dropout probability; if greater than 0.0, dropout is applied
-     is_causal (bool): If true, assumes causal attention masking and ignores attn_mask.
+    query (Tensor): Query tensor; shape (N, ..., L, E) | (N, num_heads, L, E) for fused kernels.
```

Comment:
What are the new shape annotation for? The `...` is already representing any number of dimension that match accross args/outputs for me.

Reply:
The fused kernels can only be accessed if the input shapes are 4 dimensional. I added this to try and push people down that path but I can remove

Reply:
This makes it hard to read for me. I would personally prefer an explicit message specifying that these fastpath only run in situation X Y.
But that's up to preference so you can keep what you have if you prefer it :D